content/en/blog/_posts/2021-06-08-Kubeadm-Avoid-Default-Address.md

```diff
@@ -0,0 +1,313 @@
+---
+layout: blog
+title: "Avoiding the Default Address when Using kubeadm to Create a Cluster"
+date: 2021-06-08
+slug: avoiding-the-default-address-when-using-kubeadm-to-create-a-cluster
+---
+
+**Authors:** Mike Spreitzer
+
+When using `kubeadm` to create a cluster, the default behavior is to
+take the default address of each node as the address to use in every
+sort of Kubernetes configuration --- and sometimes this is entirely
+wrong.  I am concerning myself here only with IPv4; extension to IPv6
+and dual-stack should be follow the same lines.  The default address
+of a node is the address of the network interface that appears in the
+default route.  In some use cases, this default address is not
+reachable from other nodes and so _must not_ be used in any part of
+Kuberntes configuration.  I recently had this happen to me, when I
+tried to do a quick experiment using VirtualBox VMs created in the
+most direct and simple way using the VirtualBox GUI.  I gave each VM
+two network interfaces: one using a "NAT adapter" to support
+connectivity with the external world, and one using a "Host-only
+adapter" for connectivity between the VMs.  The NAT adapter interfaces
+all get the same IP address, 10.0.2.15, and so do not provide
+connectivity between the VMs _but_ appear in the default route.  You
+may have other use cases, perhaps involving containers.
+
+In this post I will look at how to make all of the changes needed for
+a single-control-node cluster.  I have not yet attempted multiple
+control nodes, and the solution will need further work for that case
+(to give each kube-apiserver a distinct advertise-address).  The
+needed results are getting the desired address to appear in the
+following places.
```

Comment:
As far as I know, you have recited facts about Kubernetes.  If Kubernetes has an simple automated way to handle these scenarios,  please give a pointer.

Reply:
i don't think there is an automated way to achieve this.

the kubeadm "create a cluster" page has the following:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/

> (Optional) Unless otherwise specified, kubeadm uses the network interface associated with the default gateway to set the advertise address for this particular control-plane node's API server. To use a different network interface, specify the --apiserver-advertise-address=<ip-address> argument to kubeadm init. To deploy an IPv6 Kubernetes cluster using IPv6 addressing, you must specify an IPv6 address, for example --apiserver-advertise-address=fd00::101

we could add another note that indicates to users that a similar IP detection is done in all components. relevant flags of said components, already document this. if the user interface setup is "non-compliant", they would have to override a number of flags rendering this setup as not-recommended.